### PR TITLE
Improve segment file documentation

### PR DESCRIPTION
**Pull request by conradlee**
@fjy  Could you take a look at this?  I'm not sure whether I've linked to the images correctly, because I don't know how to serve the docs locally.

@xvrl Could you let me know if there are any other details you'd like to include about segment files?

### CODE REVIEW

File: docs/content/design/segments.md [hunk 1]
```diff
@@ -4,7 +4,79 @@ layout: doc_page
...
+segment file is created for each time interval, where the time inteval
+is configurable in the indexing spec. The ideal size of a segment file is
+generally 300mb-700mb, so larger data sources may need to partition their
+data into multiple segment files for each time interval--more on this below.
+
+### A segment file's core data structures
+Here we describe the internal structure of segment files, which is
+essentially *columnar*: the data for each column is laid out in
+separate data structures. By storing each column separately, Druid can
+decrease query latency by scanning only those columns actaully needed
```
Comment (fjy) [inline docs/content/design/segments.md]: actually*

**Comment by fjy** [comment]
:+1: @conradlee Can you sign the CLA at http://druid.io/community/cla.html?

File: docs/content/design/segments.md [hunk 2]
```diff
@@ -4,7 +4,79 @@ layout: doc_page
 Segments
 ========
 
-Druid segments contain data for a time interval, stored as separate columns. Dimensions (string columns) have inverted indexes associated with them for each dimension value. Metric columns are LZ4 compressed.
+
+Druid stores its index in *segment files*. In a basic setup, one
+segment file is created for each time interval, where the time inteval
+is configurable in the indexing spec. The ideal size of a segment file is
+generally 300mb-700mb, so larger data sources may need to partition their
```
Comment (nishantmonu51) [inline docs/content/design/segments.md]: also mention the typical number of rows in a segment (~5M)

File: docs/content/design/segments.md [hunk 3]
```diff
@@ -4,7 +4,79 @@ layout: doc_page
...
+the list of values in 3 are needed for *group by* and *TopN*
+queries. In other words, queries that solely aggregate metrics based
+on filters do not need to touch the list of dimension values stored in
+3.
+
+To get a concrete sense of these data structures, consider the ‘page’
+column from the example data above.  The three data structures that
+represent this dimension are illustrated in the diagram below. 
+
+![Druid Dimension Columns](../img/druid-dimension-column-details2.png "A column is represented by three data structures in a segment file")
```
Comment (xvrl) [inline docs/content/design/segments.md]: we can probably replace the image with a verbatim text block in ``` quotes

Comment (himanshug) [inline docs/content/design/segments.md]: +1 for replacing images with formatted text if/where possible.

**Comment by fjy** [comment]
@conradlee did you get a chance to look at the review comments?

**Comment by conradlee** [comment]
@fjy I took Friday and Monday off so am just now reading your comments.  I just agreed to and submitted the CLA.  I'll address the other comments now.

File: docs/content/design/segments.md [hunk 4]
```diff
@@ -4,7 +4,120 @@ layout: doc_page
...
+Druid stores its index in *segment files*. In a basic setup, one
+segment file is created for each time interval, where the time inteval
+is configurable in the indexing spec.  For druid to operate well under
+heavy query load, it is important for the segment file size to be
+within the recommended range of 300mb-700mb. If your segment files are
+larger than this range, then consider either changing the the
+granularity of the time interval or partitioning your data and
+tweaking the `targetPartitionSize` in your `partitioningSpec` (a good
+starting point for this parameter is 5 million rows).  See the section
+below on sharding for more information.
```
Comment (conradlee) [inline docs/content/design/segments.md]: Is there some way of linking to an anchor on the same page?  I think this isn't possible with regular github-style markdown (please correct me if I'm wrong), but it looks like you guys are compiling these docs with something other than github, which might allow for linking to anchors.

File: docs/content/design/segments.md [hunk 5]
```diff
@@ -4,7 +4,120 @@ layout: doc_page
...
+Druid stores its index in *segment files*. In a basic setup, one
+segment file is created for each time interval, where the time inteval
+is configurable in the indexing spec.  For druid to operate well under
+heavy query load, it is important for the segment file size to be
+within the recommended range of 300mb-700mb. If your segment files are
+larger than this range, then consider either changing the the
+granularity of the time interval or partitioning your data and
+tweaking the `targetPartitionSize` in your `partitioningSpec` (a good
+starting point for this parameter is 5 million rows).  See the section
+below on sharding or the documentation on indexing for more information.
```
Comment (conradlee) [inline docs/content/design/segments.md]: How exactly do I link to other doc files?  It looks like you guys are compiling your markdown with something other than github, and I'm not sure how the directory structure of the docs maps to the URL path structure.  I want to insert a link here to some of the batch indexing documentation that covers partitioning.

File: docs/content/design/segments.md [hunk 6]
```diff
@@ -52,4 +165,10 @@ Sharding Data to Create Segments
 
 ### Sharding Data by Dimension
```
Comment (conradlee) [inline docs/content/design/segments.md]: Is it preferred to call this 'sharding' or 'partitioning'?  I think technically sharding is a particular type of partitioning (horizontal, shared nothing partitioning), so either could work here, it's just a matter of how specific we want to be.  I think it's confusing that the documentation uses both terms---I wondered whether there these referred to different things.

File: docs/content/design/segments.md (hunk 5 above)
Comment (cheddar) [inline docs/content/design/segments.md]: it's all flattened on the github side.  If you look in the "_bin" folder, you can see the scripts that we use to generate PRs against the druid-io.github.io branch.

File: docs/content/design/segments.md (hunk 6 above)
Comment (cheddar) [inline docs/content/design/segments.md]: In general, we "partition" the data source by time and then we "shard" the partitions by some other dimension/set of dimensions.

**Pull request merged**